### PR TITLE
feat(ui): added summary-box to consolidate +x usecases

### PR DESCRIPTION
**Pull request by 23nobody**
Screenshot 2022-02-01 at 12 48 20 PM
Existing example just shows how it is being rendered, the provided dummy input is like that only.

### CODE REVIEW

File: projects/components/src/summary-box/summary-box.component.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import { ChangeDetectionStrategy, Component, Input, OnChanges } from '@angular/core';
+import { Color } from '@hypertrace/common';
+
+@Component({
+  selector: 'ht-summary-box',
```
Comment (anandtiwary) [inline projects/components/src/summary-box/summary-box.component.ts:5]: This sounds very similar to summary-value component.

File: projects/components/src/summary-box/summary-box.component.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      class="summary-text"
+      *ngIf="this.count > 0"
+      [htTooltip]="summaryTooltip"
+      [style.color]="this.color"
+      [style.backgroundColor]="this.backgroundColor"
+      [ngClass]="this.displayStyle"
+      >{{ this.summaryText }} {{ this.suffix }}</span
+    >
+    <ng-template #summaryTooltip>
+      <div class="tooltip-contents">{{ this.tooltip }}</div>
```
Comment (anandtiwary) [inline projects/components/src/summary-box/summary-box.component.ts:18]: This component only renders `<count> more` with background and tooltip. 

I am not clear how we will use it. Could you please share an example?

File: projects/components/src/summary-box/summary-box.component.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  public displayStyle: SummaryBoxDisplay = SummaryBoxDisplay.Plain;
+
+  @Input()
+  public backgroundColor?: Color = Color.White;
+
+  @Input()
+  public color?: string = Color.Gray7;
+
+  @Input()
+  public tooltip!: string;
```
Comment (anandtiwary) [inline projects/components/src/summary-box/summary-box.component.ts:40]: We already have a tooltip directive. We can use it directly on the component. No need to add it as another input. If you want, the tooltip can be an internal property for this component.

File: projects/components/src/summary-box/summary-box.component.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  public suffix?: string = 'more';
+
+  @Input()
+  public displayStyle: SummaryBoxDisplay = SummaryBoxDisplay.Plain;
+
+  @Input()
+  public backgroundColor?: Color = Color.White;
+
+  @Input()
+  public color?: string = Color.Gray7;
```
Comment (arjunlalb) [inline projects/components/src/summary-box/summary-box.component.ts:37]: we don't need individual colors if are doing a displaystyle based implementation. apply a css class based on the style that takes care of necessary styling.

File: projects/components/src/summary-box/summary-box.component.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    <span
+      class="summary-text"
+      *ngIf="this.count > 0"
+      [htTooltip]="summaryTooltip"
+      [style.color]="this.color"
+      [style.backgroundColor]="this.backgroundColor"
+      [ngClass]="this.displayStyle"
+      >{{ this.summaryText }} {{ this.suffix }}</span
+    >
+    <ng-template #summaryTooltip>
```
Comment (arjunlalb) [inline projects/components/src/summary-box/summary-box.component.ts:17]: take string | number | TemplateRef as input and supply that directly to `htTooltip`. This component shouldn't be constructing its own tooltip.

File: projects/components/src/summary-box/summary-box.component.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    </ng-template>
+  </div>`,
+  styleUrls: ['./summary-box.component.scss']
+})
+export class SummaryBoxComponent implements OnChanges {
+  @Input()
+  public count!: number;
+
+  @Input()
+  public suffix?: string = 'more';
```
Comment (arjunlalb) [inline projects/components/src/summary-box/summary-box.component.ts:28]: default to empty.

Comment (23nobody) [reply projects/components/src/summary-box/summary-box.component.ts]: done

File: projects/components/src/summary-box/summary-box.component.ts (hunk 5 above)
Comment (23nobody) [reply projects/components/src/summary-box/summary-box.component.ts]: done

File: projects/components/src/summary-box/summary-box.component.ts (hunk 4 above)
Comment (23nobody) [reply projects/components/src/summary-box/summary-box.component.ts]: didn't get you?

File: projects/components/src/summary-box/summary-box.component.ts (hunk 1 above)
Comment (23nobody) [reply projects/components/src/summary-box/summary-box.component.ts]: refactored to x-more

File: projects/components/src/summary-box/summary-box.component.ts (hunk 3 above)
Comment (23nobody) [reply projects/components/src/summary-box/summary-box.component.ts]: taking string | number | TemplateRef as input and supplying that to htTooltip.

File: projects/components/src/x-more/x-more.component.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  public count!: number;
+
+  @Input()
+  public suffix?: string = '';
+
+  @Input()
+  public displayStyle: XMoreDisplay = XMoreDisplay.Plain;
+
+  @Input()
+  public tooltip!: string | number | TemplateRef<TooltipDirective>;
```
Comment (arjunlalb) [inline projects/components/src/x-more/x-more.component.ts:25]: should be `TemplateRef<unknown>`

File: projects/components/src/x-more/x-more.component.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+  public summaryText!: string;
+
+  public ngOnChanges(): void {
+    this.summaryText = `+${this.count}`;
+
+    if (this.suffix !== '') {
+      this.summaryText += ` ${this.suffix}`;
+    }
+  }
```
Comment (arjunlalb) [inline projects/components/src/x-more/x-more.component.ts:35]: add typing to changes object and update the values only if relevant values have changed.

Comment (23nobody) [reply projects/components/src/x-more/x-more.component.ts]: done

File: projects/components/src/x-more/x-more.component.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+  public ngOnChanges(changes: TypedSimpleChanges<this>): void {
+    if (changes.count || changes.suffix) {
+      this.summaryText = `+${this.count}`;
+
+      if (this.suffix !== '') {
+        this.summaryText += ` ${this.suffix}`;
+      }
+    }
+  }
```
Comment (arjunlalb) [inline projects/components/src/x-more/x-more.component.ts:37]:
```suggestion
    if (changes.count || changes.suffix) {
      this.summaryText = this.suffix !== undefined && this.suffix !== '' ? `${this.count} ${this.suffix}` : `+${this.count}`
    }
  }
```

File: projects/components/src/x-more/x-more.component.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+  public ngOnInit(): void {
+    this.summaryText = `+${this.count}`;
+
+    if (this.suffix !== '') {
+      this.summaryText += ` ${this.suffix}`;
+    }
+  }
+}
```
Comment (arjunlalb) [inline projects/components/src/x-more/x-more.component.ts:46]: this isn't required. onChanges will take care of this.

Comment (23nobody) [reply projects/components/src/x-more/x-more.component.ts]: done

**Pull request merged**